lua/plenary/async_lib/work.lua

```diff
@@ -0,0 +1,91 @@
+local a = require('plenary.async_lib.async')
+local uv = vim.loop
+
+local M = {}
+
+local function work_map(tbl, func, callback)
+  local results = {}
+  local ran_callback = false
+
+  local work = uv.new_work(func, function(idx, ...)
+    results[idx] = {...}
+    if #results == #tbl and ran_callback == false then
+      callback(results)
+      ran_callback = true
+    end
+  end)
+
+  for idx, v in ipairs(tbl) do
+    M.validate_threadargs(v)
+
+    work:queue(idx, v)
+  end
+end
+
+M.map_async = a.wrap(work_map, 3)
+
+function M.is_threadarg(t)
+  return type(t) == "nil" or type(t) == "boolean" or type(t) == "number" or type(t) == "string" or type(t) == "userdata"
```

Comment:
only get the type once here, instead of multiple times.
also, you can minimize checks by doing:

```lua
local valid_types = {
  ["nil"] = true,
  ["boolean"] = true,
  ...
}

function threadarg(t) return valid_types[type(t)] end
```

similar or whatever, just writing this in github.

Reply:
removed for now